src/test/spec/retryable_reads.rs

```diff
@@ -0,0 +1,183 @@
+                if !data.is_empty() {
+                    coll.insert_many(data, None)
+                        .await
+                        .expect(&test_case.description);
+                }
+            }
+
+            if let Some(ref fail_point) = test_case.fail_point {
+                client
+                    .database("admin")
+                    .run_command(fail_point.clone(), None)
+                    .await
+                    .unwrap();
+            }
+
+            let mut events: Vec<TestEvent> = Vec::new();
+            for operation in test_case.operations {
+                let test_operation: AnyTestOperation =
+                    bson::from_bson(Bson::Document(operation.as_document())).unwrap();
+                let result = match operation.object.as_ref().map(String::as_ref) {
+                    Some("client") => client.run_client_operation(&test_operation).await,
+                    Some("database") => {
+                        client
+                            .run_database_operation(&test_operation, &db_name)
+                            .await
+                    }
+                    Some("collection") | None => {
+                        client
+                            .run_collection_operation(&test_operation, &db_name, &coll_name)
+                            .await
+                    }
+                    Some(op_type) => panic!("unsupported operation type {}", op_type),
+                };
+                let mut operation_events: Vec<TestEvent> = client
+                    .collect_events(&test_operation, false)
+                    .into_iter()
+                    .map(Into::into)
+                    .collect();
+
+                if let Some(error) = operation.error {
+                    if error {
+                        assert!(result.is_err());
+                    }
+                }
+                if let Some(expected_result) = operation.result {
+                    assert_eq!(result.unwrap().unwrap(), expected_result);
+                }
+
+                events.append(&mut operation_events);
+            }
+
+            if let Some(expectations) = test_case.expectations {
+                assert_eq!(events.len(), expectations.len());
+                for (actual_event, expected_event) in events.iter().zip(expectations.iter()) {
+                    assert_matches(actual_event, expected_event, None);
+                }
+            }
+
+            if test_case.fail_point.is_some() {
+                client
+                    .database("admin")
+                    .run_command(
+                        doc! {
+                            "configureFailPoint": "failCommand",
+                            "mode": "off"
+                        },
+                        None,
+                    )
+                    .await
+                    .unwrap();
+            }
+        }
+    }
+
+    run_spec_test(&["retryable-reads", "tests"], run_test).await;
```

Comment:
the spec tests can be moved up a directory so that this just has to specify "retryable-reads"